### PR TITLE
Super4PCS V1

### PR DESCRIPTION
**Pull request by nmellado**
This PR is not ready for merging yet, but instead groups major changes in the Super4PCS code structure, in order to prepare new version (v1).
Major changes are:
 - Dependencies:
     - remove ANN dependency (done),
     - remove dependency to Chealpix (done). Can be controlled by the option `SUPER4PCS_USE_CHEALPIX`. This options will be available for some time for experiment purposes. During this time, Chealpix and Cfitsio files will still be shipped with Super4PCS source code. 
     - OpenCV is required only to load meshes textures. This dependency can be removed by switching option `IO_USE_OPENCV` to `FALSE`,
 - remove duplicate code between 4PCS and Super4PCS (done),
 - clean code structure,
 - demo dataset + run script added

Consequences of this changes:
 - better maintainability
 - improved testing capabilities
 - better performances
 - compilation made easier (cross-plateform, compile-time, etc.)

Tested on:
 - Debian 4.9.30-1 
   - clang version 3.8.1-24
   - gcc version 6.3.0 20170516
 - Microsoft Windows 10
    - VS 2017

### CODE REVIEW

File: src/super4pcs/accelerators/pairExtraction/bruteForceFunctor.h [hunk 1]
```diff
@@ -87,37 +89,36 @@ struct BruteForceFunctor{
...
     )
 {
-  // This functor is optimized to be ran on a single point set
-  // \FIXME Should implement a better way to deal with that...
-  assert(M.size() != Q.size());
 
-  const unsigned int size = Q.size();
+  const unsigned int sizeM = M.size();
+  const unsigned int sizeQ = Q.size();
   unsigned int i, j;
```
Comment (simogasp) [inline src/super4pcs/accelerators/pairExtraction/bruteForceFunctor.h]: you can just declare i and j in the for statement

File: src/super4pcs/accelerators/pairExtraction/intersectionFunctor.h [hunk 2]
```diff
@@ -136,79 +143,77 @@ IntersectionFunctor<Primitive, Point, dim, Scalar>::process(
...
-          // current node, in that case we stop splitting it, or we split.          
+          // current node, in that case we stop splitting it, or we split.
           if (n.rangeLength() > int(minNodeSize)){
-//#pragma omp critical  
-            n.split(*childNodes, edgeHalfLength);                    
+//#pragma omp critical
+            n.split(*childNodes, edgeHalfLength);
           }else{
 //#pragma omp critical
             earlyNodes.push_back(std::pair<Node, Scalar>(n, edgeHalfLength+epsilon));
```
Comment (simogasp) [inline src/super4pcs/accelerators/pairExtraction/intersectionFunctor.h]: that could be written as
```
earlyNodes.emplace_back(n, edgeHalfLength+epsilon);
```

File: src/super4pcs/algorithms/4pcs.cc [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+// distances.
+bool Match4PCS::FindCongruentQuadrilaterals(
+        Scalar invariant1,
+        Scalar invariant2,
+        Scalar /*distance_threshold1*/,
+        Scalar distance_threshold2,
+        const std::vector<std::pair<int, int>>& P_pairs,
+        const std::vector<std::pair<int, int>>& Q_pairs,
+        std::vector<Super4PCS::Quadrilateral>* quadrilaterals) const {
+  if (quadrilaterals == NULL) return false;
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: `== nullptr`

File: src/super4pcs/algorithms/4pcs.cc [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    data_points[i * 2 + 1][0] = p1.x + (1.0-invariant1) * (p2.x - p1.x);
+    data_points[i * 2 + 1][1] = p1.y + (1.0-invariant1) * (p2.y - p1.y);
+    data_points[i * 2 + 1][2] = p1.z + (1.0-invariant1) * (p2.z - p1.z);
+  }
+
+  ANNkd_tree* tree = new ANNkd_tree(data_points, number_of_points, 3);
+
+    //Point3D invRes;
+  // Query the ANN for all the points corresponding to the invariants in Q_pair.
+  for (int i = 0; i < Q_pairs.size(); ++i) {
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: better use `std::size_t i`

File: src/super4pcs/algorithms/4pcs.cc [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        Scalar invariant1,
+        Scalar invariant2,
+        Scalar /*distance_threshold1*/,
+        Scalar distance_threshold2,
+        const std::vector<std::pair<int, int>>& P_pairs,
+        const std::vector<std::pair<int, int>>& Q_pairs,
+        std::vector<Super4PCS::Quadrilateral>* quadrilaterals) const {
+  if (quadrilaterals == NULL) return false;
+
+  int number_of_points = 2 * P_pairs.size();
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: better use `std::size_t number_of_points`

File: src/super4pcs/algorithms/4pcs.cc [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+    // We test the other order as our pairs are not ordered.
+    query_point[0] = p1.x + (1.0-invariant2) * (p2.x - p1.x);
+    query_point[1] = p1.y + (1.0-invariant2) * (p2.y - p1.y);
+    query_point[2] = p1.z + (1.0-invariant2) * (p2.z - p1.z);
+
+    tree->annkFRSearch(query_point, distance_threshold2, number_of_points,
+                       near_neighbor_index, distances, 0);
+
+    for (int j = 0; j < number_of_points; ++j) {
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: std::size_t

File: src/super4pcs/algorithms/4pcs.cc [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+// in P, by having the same distance (up to some tolerantz) and optionally the
+// same angle between normals and same color.
+void
+Match4PCS::ExtractPairs(Scalar pair_distance,
+                        Scalar pair_normals_angle,
+                        Scalar pair_distance_epsilon,
+                        int base_point1,
+                        int base_point2,
+                        PairsVector* pairs) const {
+  if (pairs == NULL) return;
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: nullptr

File: src/super4pcs/algorithms/4pcs.cc [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+  pairs->reserve(2 * sampled_Q_3D_.size());
+
+  cv::Point3d segment1 = base_3D_[base_point2] - base_3D_[base_point1];
+  segment1 *= 1.0 / cv::norm(segment1);
+
+  const Scalar norm_threshold =
+      0.5 * options_.max_normal_difference * M_PI / 180.0;
+
+  // Go over all ordered pairs in Q.
+  for (int j = 0; j < sampled_Q_3D_.size(); ++j) {
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: size_t

File: src/super4pcs/algorithms/4pcs.cc [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    const Point3D& p = sampled_Q_3D_[j];
+    for (int i = j + 1; i < sampled_Q_3D_.size(); ++i) {
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: size_t

File: src/super4pcs/algorithms/4pcs.cc [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        // Add ordered pair.
+        pairs->push_back(std::make_pair(j, i));
+      }
+      // The same for the second order.
+      segment2 = p - q;
+      segment2 *= 1.0 / cv::norm(segment2);
+      if (acos(segment1.dot(segment2)) <= options_.max_angle * M_PI / 180.0 &&
+          dist_good && rgb_good) {
+        // Add ordered pair.
+        pairs->push_back(std::make_pair(i, j));
```
Comment (simogasp) [inline src/super4pcs/algorithms/4pcs.cc]: `emplace_back(i, j)`

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 11]
```diff
@@ -0,0 +1,951 @@
...
+    // call Virtual handler
+    Initialize(P, Q);
+
+    best_LCP_ = Verify(transform_);
+    printf("Initial LCP: %f\n", best_LCP_);
+}
+
+
+bool Match4PCSBase::SelectRandomTriangle(int* base1, int* base2, int* base3) {
+      if (base1 == NULL || base2 == NULL || base3 == NULL) return false;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: nullptr

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 12]
```diff
@@ -0,0 +1,951 @@
...
+      *base1 = *base2 = *base3 = -1;
+
+      // Pick the first point at random.
+      int first_point = rand() % number_of_points;
+
+      // Try fixed number of times retaining the best other two.
+      Scalar best_wide = 0.0;
+      for (int i = 0; i < kNumberOfDiameterTrials; ++i) {
+        // Pick and compute
+        int second_point = rand() % number_of_points;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: these four can be all `const`

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 13]
```diff
@@ -0,0 +1,951 @@
...
+        Scalar how_wide = cv::norm(u.cross(w));
+        if (how_wide > best_wide && cv::norm(u) < max_base_diameter_ &&
+            cv::norm(w) < max_base_diameter_) {
+          best_wide = how_wide;
+          *base1 = first_point;
+          *base2 = second_point;
+          *base3 = third_point;
+        }
+      }
+      if (*base1 == -1 || *base2 == -1 || *base3 == -1)
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: `return (*base1 != -1 && *base2 != -1 && *base3 != -1)` ;-)

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 14]
```diff
@@ -0,0 +1,951 @@
...
+}
+
+
+
+// Try the current base in P and obtain the best pairing, i.e. the one that
+// gives the smaller distance between the two closest points. The invariants
+// corresponding the the base pairing are computed.
+bool Match4PCSBase::TryQuadrilateral(Scalar *invariant1, Scalar *invariant2,
+                                     int& id1, int& id2, int& id3, int& id4) {
+  if (invariant1 == NULL || invariant2 == NULL) return false;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: nullptr

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 15]
```diff
@@ -0,0 +1,951 @@
...
+
+  float min_distance = FLT_MAX;
+  int best1, best2, best3, best4;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: it's better to initialize them, to -1 maybe.
and at the end of the loop put some asserts just to be sure.

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 16]
```diff
@@ -0,0 +1,951 @@
...
+  return true;
+}
+
+
+// Selects a good base from P and computes its invariants. Returns false if
+// a good planar base cannot can be found.
+bool Match4PCSBase::SelectQuadrilateral(Scalar* invariant1, Scalar* invariant2,
+                                        int* base1, int* base2, int* base3,
+                                        int* base4) {
+  if (invariant1 == NULL || invariant2 == NULL || base1 == NULL ||
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: nullptr

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 17]
```diff
@@ -0,0 +1,951 @@
...
+          (-y2 * z1 + y3 * z1 + y1 * z2 - y3 * z2 - y1 * z3 + y2 * z3) / denom;
+      double B =
+          (x2 * z1 - x3 * z1 - x1 * z2 + x3 * z2 + x1 * z3 - x2 * z3) / denom;
+      double C =
+          (-x2 * y1 + x3 * y1 + x1 * y2 - x3 * y2 - x1 * y3 + x2 * y3) / denom;
+      *base4 = -1;
+      double best_distance = FLT_MAX;
+      // Go over all points in P.
+      for (unsigned int i = 0; i < sampled_P_3D_.size(); ++i) {
+        double d1 = PointsDistance(sampled_P_3D_[i], sampled_P_3D_[*base1]);
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: all the 4 variables can be `const`

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 18]
```diff
@@ -0,0 +1,951 @@
...
+        double d2 = PointsDistance(sampled_P_3D_[i], sampled_P_3D_[*base2]);
+        double d3 = PointsDistance(sampled_P_3D_[i], sampled_P_3D_[*base3]);
+        float too_small = max_base_diameter_ * kBaseTooSmall;
+        if (d1 >= too_small && d2 >= too_small && d3 >= too_small) {
+          // Not too close to any of the first 3.
+          double distance =
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: `const`

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 19]
```diff
@@ -0,0 +1,951 @@
...
+    }
+    current_trial++;
+  }
+
+  // We failed to find good enough base..
+  return false;
+}
+
+void Match4PCSBase::initKdTree(){
+  int number_of_points = sampled_P_3D_.size();
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: std::size_t

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 20]
```diff
@@ -0,0 +1,951 @@
...
+
+  // Build the kdtree.
+  kd_tree_ = Super4PCS::KdTree<Scalar>(number_of_points);
+
+  Super4PCS::KdTree<Scalar>::VectorType p;
+  for (int i = 0; i < number_of_points; ++i) {
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: std::size_t

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 21]
```diff
@@ -0,0 +1,951 @@
...
+
+    // set the basis coordinates in the congruent quad array
+    congruent_points[0].first = b1;
+    congruent_points[1].first = b2;
+    congruent_points[2].first = b3;
+    congruent_points[3].first = b4;
+
+    Eigen::Matrix<Scalar, 4, 4> transform;
+    for (int i = 0; i < congruent_quads.size(); ++i) {
+      int a = congruent_quads[i].vertices[0];
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: const for all 4

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 22]
```diff
@@ -0,0 +1,951 @@
...
+                << d << std::endl;
+  #endif
+
+      centroid2 << (congruent_points[0].second.x + congruent_points[1].second.x + congruent_points[2].second.x) / Scalar(3.),
+                   (congruent_points[0].second.y + congruent_points[1].second.y + congruent_points[2].second.y) / Scalar(3.),
+                   (congruent_points[0].second.z + congruent_points[1].second.z + congruent_points[2].second.z) / Scalar(3.);
+
+      Scalar rms = -1;
+
+      bool ok =
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: const

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 23]
```diff
@@ -0,0 +1,951 @@
...
+        cv::Point3d  q2 = pairs[2].second;
+
+  Scalar scaleEst (1.);
+
+  // Compute scale factor if needed
+  if (computeScale){
+      const cv::Point3d& p3 = pairs[3].first;
+      const cv::Point3d& q3 = pairs[3].second;
+
+      Scalar ratio1 = cv::norm(p1 - p0) / cv::norm(q1 - q0);
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: `const` for all the 4 vars

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 24]
```diff
@@ -0,0 +1,951 @@
...
+  // Discard singular solutions. The rotation should be orthogonal.
+  cv::Mat unit = rotation * rotation.t();
+  if (std::abs(unit.at<double>(0, 0) - 1.0) > kSmallNumber ||
+      std::abs(unit.at<double>(1, 1) - 1.0) > kSmallNumber ||
+      std::abs(unit.at<double>(2, 2) - 1.0) > kSmallNumber){
+      return false;
+  }
+
+  // Discard too large solutions (todo: lazy evaluation during boolean computation
+  Scalar theta_x = std::abs(std::atan2(rotation.at<double>(2, 1), rotation.at<double>(2, 2)));
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: const for all of 3

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 25]
```diff
@@ -0,0 +1,951 @@
...
+// early termination. It was found to be fast in practice.
+Match4PCSBase::Scalar
+Match4PCSBase::Verify(const Eigen::Matrix<Scalar, 4, 4>& mat) {
+
+#ifdef TEST_GLOBAL_TIMINGS
+    Timer t_verify (true);
+#endif
+
+  // We allow factor 2 scaling in the normalization.
+  Scalar epsilon = options_.delta;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: const

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 26]
```diff
@@ -0,0 +1,951 @@
...
+  int good_points = 0;
+  int number_of_points = sampled_Q_3D_.size();
+  int terminate_value = best_LCP_ * number_of_points;
+
+  Scalar sq_eps = epsilon*epsilon;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: const

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 27]
```diff
@@ -0,0 +1,951 @@
...
+
+  return best_LCP_;
+}
+
+
+// Performs N RANSAC iterations and compute the best transformation. Also,
+// transforms the set Q by this optimal transformation.
+bool Match4PCSBase::Perform_N_steps(int n, cv::Mat* transformation,
+                                    std::vector<Point3D>* Q) {
+  if (transformation == NULL || Q == NULL) return false;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: nullptr

File: src/super4pcs/algorithms/match4pcsBase.cc [hunk 28]
```diff
@@ -0,0 +1,951 @@
...
+
+#ifdef TEST_GLOBAL_TIMINGS
+    Timer t (true);
+#endif
+
+  Scalar last_best_LCP = best_LCP_;
+  bool ok;
```
Comment (simogasp) [inline src/super4pcs/algorithms/match4pcsBase.cc]: init to false?

File: src/super4pcs/algorithms/super4pcs.cc [hunk 29]
```diff
@@ -0,0 +1,247 @@
...
+        Scalar invariant1,
+        Scalar invariant2,
+        Scalar distance_threshold1,
+        Scalar distance_threshold2,
+        const std::vector<std::pair<int, int>>& P_pairs,
+        const std::vector<std::pair<int, int>>& Q_pairs,
+        std::vector<Super4PCS::Quadrilateral>* quadrilaterals) const {
+
+
+  if (quadrilaterals == NULL) return false;
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: nullptr

File: src/super4pcs/algorithms/super4pcs.cc [hunk 30]
```diff
@@ -0,0 +1,247 @@
...
+
+  quadrilaterals->clear();
+
+  // Compute the angle formed by the two vectors of the basis
+  Point3D b1 = base_3D_[1] - base_3D_[0];  b1.normalize();
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: cosnt

File: src/super4pcs/algorithms/super4pcs.cc [hunk 31]
```diff
@@ -0,0 +1,247 @@
...
+//                    Scalar>  //! \brief Scalar type
+//  IndexedNormalSet3D;
+
+
+  IndexedNormalSet3D nset (eps, 2.);
+
+  for (unsigned int i = 0; i <  P_pairs.size(); ++i) {
+    const Point& p1 = pcfunctor_.points[P_pairs[i].first];
+    const Point& p2 = pcfunctor_.points[P_pairs[i].second];
+    Point  n  = (p2 - p1).normalized();
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: const

File: src/super4pcs/algorithms/super4pcs.cc [hunk 32]
```diff
@@ -0,0 +1,247 @@
...
+  for (unsigned int i = 0; i < Q_pairs.size(); ++i) {
+    const Point& p1 = pcfunctor_.points[Q_pairs[i].first];
+    const Point& p2 = pcfunctor_.points[Q_pairs[i].second];
+
+    const Point3D& pq1 = sampled_Q_3D_[Q_pairs[i].first];
+    const Point3D& pq2 = sampled_Q_3D_[Q_pairs[i].second];
+
+    nei.clear();
+
+    Point   query  =  p1 + invariant2 * ( p2 - p1 );
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: const

File: src/super4pcs/algorithms/super4pcs.cc [hunk 33]
```diff
@@ -0,0 +1,247 @@
...
+    Point3D queryQ = pq1 + invariant2 * (pq2 - pq1);
+
+    Point queryn = (p2 - p1).normalized();
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: const

File: src/super4pcs/algorithms/super4pcs.cc [hunk 34]
```diff
@@ -0,0 +1,247 @@
...
+//         << query.transpose() << endl
+//         << queryn.transpose() << endl;
+
+    nset.getNeighbors( query,  queryn, alpha, nei);
+
+
+    Point3D invPoint;
+    //const float distance_threshold2s = distance_threshold2 * distance_threshold2;
+    for (unsigned int k = 0; k != nei.size(); k++){
+      int id = nei[k];
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: const

File: src/super4pcs/algorithms/super4pcs.cc [hunk 35]
```diff
@@ -0,0 +1,247 @@
...
+
+      const Point3D& pp1 = sampled_Q_3D_[P_pairs[id].first];
+      const Point3D& pp2 = sampled_Q_3D_[P_pairs[id].second];
+
+      invPoint = pp1 + (pp2 - pp1) * invariant1;
+
+       // use also distance_threshold2 for inv 1 and 2 in 4PCS
+      if (cv::norm(queryQ-invPoint) <= distance_threshold2){
+          comb.insert(std::make_pair(id, i));
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: `comb.emplace(id,i)`

File: src/super4pcs/algorithms/super4pcs.cc [hunk 36]
```diff
@@ -0,0 +1,247 @@
...
+    }
+  }
+
+  for (std::set< std::pair<unsigned int, unsigned int > >::const_iterator it =
+             comb.cbegin();
+       it != comb.cend(); it++){
+    const unsigned int & id = (*it).first;
+    const unsigned int & i  = (*it).second;
+
+    quadrilaterals->push_back(
```
Comment (simogasp) [inline src/super4pcs/algorithms/super4pcs.cc]: `.emplace_back(P_pairs[id].first, P_pairs[id].second, Q_pairs[i].first,  Q_pairs[i].second)`

File: src/super4pcs/shared4pcs.h [hunk 37]
```diff
@@ -138,7 +137,86 @@ class Point3D : public cv::Point3f {
...
+// invariants corresponding to the closet points. This is the "intersection"
+// point that determines the invariants. Since the 4 points are not exactly
+// planar, we use the center of the line segment connecting the two closest
+// points as the "intersection".
+static float distSegmentToSegment(const cv::Point3d& p1, const cv::Point3d& p2,
+                           const cv::Point3d& q1, const cv::Point3d& q2,
+                           double* invariant1, double* invariant2) {
+  if (invariant1 == 0 || invariant2 == 0) return kLargeNumber;
+  const float kSmallNumber = 0.0001;
+  cv::Point3d u = p2 - p1;
```
Comment (simogasp) [inline src/super4pcs/shared4pcs.h]: all const

File: src/super4pcs/shared4pcs.h [hunk 38]
```diff
@@ -138,7 +137,86 @@ class Point3D : public cv::Point3f {
...
+    else if ((-d + b) > a)
+      s1 = s2;
+    else {
+      s1 = (-d + b);
+      s2 = a;
+    }
+  }
+  *invariant1 = (abs(s1) < kSmallNumber ? 0.0 : s1 / s2);
+  *invariant2 = (abs(t1) < kSmallNumber ? 0.0 : t1 / t2);
+  cv::Point3d distance = w + (*invariant1 * u) - (*invariant2 * v);
```
Comment (simogasp) [inline src/super4pcs/shared4pcs.h]: const

File: tests/pair_extraction.cc [hunk 39]
```diff
@@ -73,122 +75,133 @@
...
-    Point p (0.5f*Point::Random() + half);
+  std::vector<Primitive> primitives;
 
+  //std::cout << "Points: " << std::endl;
+  Point half (Point::Ones()/2.f);
+  for(unsigned int i = 0; i != nbPoints; i++){
+    // generate points on a sphere
+    Point p (0.5f*Point::Random().normalized() + half);
+    points.push_back(p);
     primitives.push_back(Primitive(p, r));
```
Comment (simogasp) [inline tests/pair_extraction.cc]: emplace_back(p,r)

**Comment by nmellado** [comment]
Thanks for the review @simogasp.
I've mostly focused on the code structure for now, but these changes make sens, I'll take them into account during the upcoming cleaning pass.
Thanks !

File: src/super4pcs/algorithms/super4pcs.cc (hunk 30 above)
Comment (nmellado) [inline src/super4pcs/algorithms/super4pcs.cc]: Cannot be const, call to normalize discards qualifiers

**Comment by nmellado** [comment]
@simogasp yes sure, thanks... I've been too lazy here.

File: src/super4pcs/accelerators/kdtree.h [hunk 40]
```diff
@@ -343,7 +343,7 @@ KdTree<Scalar, Index>::finalize()
 {
     mNodes.clear();
     mNodes.reserve(4*mPoints.size()/_nofPointsPerCell);
-    mNodes.push_back(KdNode());
+    mNodes.emplace_back(KdNode());
```
Comment (simogasp) [inline src/super4pcs/accelerators/kdtree.h]: mNodes.emplace_back()

File: src/super4pcs/accelerators/pairExtraction/intersectionFunctor.h [hunk 41]
```diff
@@ -140,12 +140,12 @@ IntersectionFunctor<Primitive, Point, dim, Scalar>::process(
...
     functor.ids.clear();
     for(unsigned int i = 0; i < nbPoint; i++)
-      functor.ids.push_back(i);
+      functor.ids.emplace_back(i);
   }
 
 
   // Buid root node in the child node, will be copied to the current nodes
-  childNodes->push_back(Node::buildUnitRootNode(Q, functor.ids));
+  childNodes->emplace_back(Node::buildUnitRootNode(Q, functor.ids));
```
Comment (simogasp) [inline src/super4pcs/accelerators/pairExtraction/intersectionFunctor.h]: emplace_back(Q, functor.ids)

File: src/super4pcs/io/io_ply.h [hunk 42]
```diff
@@ -188,15 +188,15 @@ readBinary1Body (const std::string & filename,
             fread (v, 4, numOfVertexProperties, in);
         if (bigEndian == true)
             bigLittleEndianSwap (v, numOfVertexProperties);
-        vertex.push_back( Point3D(v[0],v[1],v[2]) );
+        vertex.emplace_back( Point3D(v[0],v[1],v[2]) );
```
Comment (simogasp) [inline src/super4pcs/io/io_ply.h]: emplace_back(v[0],v[1],v[2])

File: src/super4pcs/io/io_ply.h [hunk 43]
```diff
@@ -228,7 +228,7 @@ readBinary1Body (const std::string & filename,
             /*count = */fread (f, 4, 3, in);
             if (bigEndian == true)
                 bigLittleEndianSwap (f, 3);
-            face.push_back(tripple(f[0],f[1],f[2]));
+            face.emplace_back(tripple(f[0],f[1],f[2]));
```
Comment (simogasp) [inline src/super4pcs/io/io_ply.h]: emplace_back(f[0],f[1],f[2])

File: src/super4pcs/io/io_ply.h [hunk 44]
```diff
@@ -293,8 +293,8 @@ readASCII1Body (const std::string & filename,
         else
             for (unsigned int j = 0;  j < numOfVertexProperties;  j++)
                 fscanf (in, "%f", &v[j]);
-        
-        vertex.push_back( Point3D(v[0],v[1],v[2]) );
+
+        vertex.emplace_back( Point3D(v[0],v[1],v[2]) );
```
Comment (simogasp) [inline src/super4pcs/io/io_ply.h]: emplace_back(v[0],v[1],v[2])

File: src/super4pcs/io/io_ply.h [hunk 45]
```diff
@@ -331,7 +331,7 @@ readASCII1Body (const std::string & filename,
             int f[3];
             int polygonSize;
             /*count = */fscanf (in, "%d %d %d %d", &polygonSize, &f[0], &f[1], &f[2]);
-            face.push_back(tripple(f[0],f[1],f[2]));
+            face.emplace_back(tripple(f[0],f[1],f[2]));
```
Comment (simogasp) [inline src/super4pcs/io/io_ply.h]: emplace_back(f[0],f[1],f[2])

File: src/super4pcs/pairCreationFunctor.h [hunk 46]
```diff
@@ -118,8 +118,8 @@ struct PairCreationFunctor{
     for (unsigned int i = 0; i < nSamples; ++i) {
       points[i] = worldToUnit(points[i]);
 
-      primitives.push_back(Primitive(points[i], Scalar(1.)));
-      ids.push_back(i);
+      primitives.emplace_back(Primitive(points[i], Scalar(1.)));
```
Comment (simogasp) [inline src/super4pcs/pairCreationFunctor.h]: emplace_back(points[i], Scalar(1.))

File: tests/pair_extraction.cc [hunk 47]
```diff
@@ -136,8 +136,8 @@ void testFunction( Scalar r, Scalar epsilon,
   for(unsigned int i = 0; i != nbPoints; i++){
     // generate points on a sphere
     Point p (0.5f*Point::Random().normalized() + half);
-    points.push_back(p);
-    primitives.push_back(Primitive(p, r));
+    points.emplace_back(p);
+    primitives.emplace_back(Primitive(p, r));
```
Comment (simogasp) [inline tests/pair_extraction.cc]: emplace_back(p, r)

File: tests/pair_extraction.cc [hunk 48]
```diff
@@ -169,7 +169,7 @@ void testFunction( Scalar r, Scalar epsilon,
     for(unsigned int i = 0; i != nbPoints; i++)
       for(unsigned int j = i+1; j < nbPoints; j++)
          if (primitives[j].intersectPoint(points[i], epsilon))
-          p2.push_back(std::make_pair(i,j));
+          p2.emplace_back(std::make_pair(i,j));
```
Comment (simogasp) [inline tests/pair_extraction.cc]: emplace_back(i,j)

File: tests/pair_extraction.cc [hunk 49]
```diff
@@ -281,8 +281,8 @@ void callMatchSubTests()
                 const Point3D& q = Q[i];
                 const Scalar distance = cv::norm(q - p);
                 if (std::abs(distance - pair_distance) <= pair_distance_epsilon) {
-                    pairs.push_back(std::make_pair(j, i));
-                    pairs.push_back(std::make_pair(i, j));
+                    pairs.emplace_back(std::make_pair(j, i));
```
Comment (simogasp) [inline tests/pair_extraction.cc]: emplace_back(j, i)

File: tests/pair_extraction.cc [hunk 50]
```diff
@@ -281,8 +281,8 @@ void callMatchSubTests()
...
+                    pairs.emplace_back(std::make_pair(i, j));
```
Comment (simogasp) [inline tests/pair_extraction.cc]: emplace_back(i,j)

**Comment by nmellado** [comment]
Thanks for the resource on clang-tidy, I'll have a look at it once I'm done with bugfixes and dependencies removal.

**Pull request merged**